### ISSUE TITLE
Question about disk layout limitations

### ISSUE THREAD

**Issue by jsmeix** [rear/rear#2023]
I would like to know what the limitations of disk layout complexity are
where the disk layout can be recreated with our current layout code.

For what I mean with "disk layout complexity" have a look at the section
"Changes in the Partitioner UI to Unleash the Storage-ng Power" in
https://lizards.opensuse.org/2018/10/09/yast-sprint-64/

The image therein
![image](https://user-images.githubusercontent.com/1788608/51597951-c9569f80-1efc-11e9-90a1-04e6b425cad6.png)
shows an (intentional artificial) example of a complex disk layout
that can be created in newest SLES15 with the so called "storage-ng"
partitioner and the storage setup subsystem in YaST.

It contains in particular things like
* MD RAID that consists of full disks and partitions (e.g. /dev/md1 consists of /dev/sdb and /dev/sda3)
* MD RAID that contains partitions (e.g. /dev/md1 contains /dev/md1p1 and /dev/md1p2)
* A disk that is directly formatted with no partitions in between (e.g. there is an ext4 filesystem directly on /dev/sdc that is mounted at /data)

That image shows an example of a higher stack of storage objects
where one kind of storage object (partitions) appear on different levels:
* mountpoint /home is on top of
* filesystem ext4 that is on top of
* partition /dev/md1p1 that is on top of
* MD RAID /dev/md1 that is on top of two parents
* disk /dev/sdb (that has no parent) and partition /dev/sda3 that is on top of
* disk /dev/sda (that has no parent)

Simply put, my question is if ReaR's current layout code is meant
to support arbitrary big mesh-like stacks of storage objects
('mesh-like' because a child can have more than one parent
and 'stack' because there is a hierarchical top-down structure)
or if there are currenly "built-in" limitations.

**Comment by jsmeix** [comment]
@schlomo @gdha 
I hope you could answer my question.
Perhaps you already experienced generic limitations in our current layout code?

**Comment by jhoekx** [comment]
The layout code was designed to handle the following setup (no picture as nice as yours):
```
HP RAID controller
  /dev/sda
    /dev/sda1
      /boot
    /dev/sda2
      pv
        lv /
        lv ...
        drbd
          pv
            lv /data1
            lv /data2
```

This means that it should be able to handle arbitrary setups as long as the names of the things are unique. The dependencies between components are used to create the devices in the right order.

**Comment by jsmeix** [comment]
@jhoekx 
thank you very much for your prompt reply.

Now I will try out (as time permits) to set up such kind of
(intentional artificial) examples of complex disk layout
and see how it works with ReaR.

**Comment by schlomo** [comment]
@jsmeix IMHO ReaR should support that complexity as I don't see there any new block device types.

If this "storage-ng" introduces a new storage layer with new block devices then we would have to add support for this.

The best way to help ReaR in this context is to create an automated test case that runs for example in a qemu VM.

**Comment by jsmeix** [comment]
Independent of the disk layout complexity (i.e. the dependencies
of parents and children in the mesh-like stack of storage objects)
there are limitations what kind of storage objects ReaR can deal with.

Currently ReaR does not support

* One Btrfs setup on multiple disks,
cf. https://github.com/rear/rear/issues/2028

* Bcache storage objects e.g. `/dev/bcache0`,
cf. https://github.com/rear/rear/issues/2057

**Comment by jsmeix** [comment]
https://github.com/rear/rear/issues/2087#issuecomment-477551079
shows an example where the current layout code failed
to get the right depencencies between storage objects.

**Comment by jsmeix** [comment]
My question was sufficiently answered.

**Comment by jsmeix** [comment]
By the way:
My added initial "barrel" support in ReaR 2.8
via https://github.com/rear/rear/issues/2863
and https://github.com/rear/rear/pull/2876
is meant to solve that ReaR can recreate (by using "barrel")
whatever disk layout SLES can setup with libstorage/storage-ng.
